econml/cate_estimator.py

```diff
@@ -436,6 +440,93 @@ def const_marginal_effect_inference(self, X=None):
+        if treatment_names is None:
+            treatment_names = [f"T{i}" for i in range(d_t)]
+        if output_names is None:
+            output_names = [f"Y{i}" for i in range(d_y)]
+
+        return _shap_explain_cme(self.const_marginal_effect, X, d_y, d_t, feature_names, treatment_names, output_names)
+
+    def _shap_values(self, models, X, *, feature_names=None, treatment_names=None, output_names=None):
+        """ Shap value for the final stage models.
+        This is the private method mainly for internal use.
+
+        Parameters
+        ----------
+        models: a single estimator or a list of estimators with one estimator per treatment
+            models for the model's final stage model.
+        X: (m, d_x) matrix
+            Features for each sample. Should be in the same shape of fitted X in final stage.
+        feature_names: optional None or list of strings of length X.shape[1] (Default=None)
+            The names of input features.
+        treatment_names: optional None or list (Default=None)
+            The name of treatment. In discrete treatment scenario, the name should not include control name.
+        output_names:  optional None or list (Default=None)
+            The name of the outcome.
+
+        Returns
+        -------
+        shap_outs: nested dictionary of Explanation object
+            A nested dictionary by using each Y and each T as key and the shap_values explanation object as value.
+        """
+
+        d_t = self._d_t[0] if self._d_t else 1
+        d_y = self._d_y[0] if self._d_y else 1
+        if treatment_names is None:
+            treatment_names = [f"T{i}" for i in range(d_t)]
+        if output_names is None:
+            output_names = [f"Y{i}" for i in range(d_y)]
+
+        if not isinstance(models, list):
+            models = [models]
+        assert len(models) == d_t, "Number of final stage models don't equals to number of treatments!"
+        # define masker by using entire dataset, otherwise Explainer will only sample 100 obs by default.
+        background = shap.maskers.Independent(X, max_samples=X.shape[0])
+
+        shap_outs = defaultdict(dict)
+        for i in range(d_t):
+            try:
+                explainer = shap.Explainer(models[i], background,
+                                           feature_names=feature_names)
+            except Exception as e:
+                print("Final model can't be parsed, explain const_marginal_effect() instead!")
+                return _shap_explain_cme(self.const_marginal_effect, X, d_y, d_t, feature_names, treatment_names,
+                                         output_names)
+            shap_out = explainer(X)
+            if d_y > 1:
+                for j in range(d_y):
+                    shap_out_copy = deepcopy(shap_out)
+                    shap_out_copy.base_values = shap_out_copy.base_values[..., j]
+                    shap_out_copy.values = shap_out_copy.values[..., j]
```

Comment:
Good Point! I just checked slicing, it could only slice a specific row or column, but in this case here, we only slice the `.base_values` and `.values`, `.data` is same for each dimension of Y.